src/main/java/com/rackspacecloud/metrics/ingestionservice/influxdb/GCLineProtocolBackupService.java

```diff
@@ -0,0 +1,133 @@
+package com.rackspacecloud.metrics.ingestionservice.influxdb;
+
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.rackspacecloud.metrics.ingestionservice.influxdb.config.BackupProperties;
+import com.rackspacecloud.metrics.ingestionservice.influxdb.providers.LineProtocolBackupService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cache.annotation.CacheEvict;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.stereotype.Service;
+import org.springframework.util.Assert;
+
+import javax.annotation.PreDestroy;
+import javax.annotation.Resource;
+import java.io.IOException;
+import java.nio.channels.Channels;
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.util.UUID;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.zip.GZIPOutputStream;
+
+@Service
+@EnableConfigurationProperties(BackupProperties.class)
+public class GCLineProtocolBackupService implements LineProtocolBackupService {
+
+    // Reusable thread-safe date-time formatter for files going in a bucket
+    private static final DateTimeFormatter dateBucketFormat = DateTimeFormatter.ofPattern("yyyyMMdd")
+            .withZone(ZoneOffset.UTC);
+    private static final Logger LOGGER = LoggerFactory.getLogger(GCLineProtocolBackupService.class);
+    // The underlying blob storage. Dev/test will uses in-memory storage
+    private final Storage storage;
+    // The bucket where blobs will go
+    private final String cloudOutputBucket;
+
+    // Internal methods of this class attempt to talk to the Cache proxy
+    // instead of the GCLineProtocolBackupService object
+    @Resource
+    private LineProtocolBackupService self;
+
+    // To autowire Storage in prod we'll need to get an account and point
+    // GOOGLE_APPLICATION_CREDENTIALS to the json creds file
+    // for a service account
+    // as per https://www.baeldung.com/java-google-cloud-storage
+    @Autowired
+    public GCLineProtocolBackupService(Storage storage, BackupProperties backupProperties) {
+        Assert.notNull(storage, "Storage must not be null");
```

Comment:
Is it possible to throw `IllegalArgumentException` instead of using `Assert`...

Reply:
It is confusing, but those are not Java's `assert` that method is Spring's and does throw `IllegalArgumentException` is much more readable way:

https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/util/Assert.html#notNull-java.lang.Object-java.lang.String-